pkg/controllers/csi/driver/volumes/app/publisher.go

```diff
@@ -22,163 +22,166 @@ import (
 	"io"
 	"os"
 	"path/filepath"
-	"strings"
+	"time"
 
-	dtcsi "github.com/Dynatrace/dynatrace-operator/pkg/controllers/csi"
 	csivolumes "github.com/Dynatrace/dynatrace-operator/pkg/controllers/csi/driver/volumes"
 	"github.com/Dynatrace/dynatrace-operator/pkg/controllers/csi/metadata"
+	"github.com/Dynatrace/dynatrace-operator/pkg/util/timeprovider"
 	"github.com/container-storage-interface/spec/lib/go/csi"
 	"github.com/pkg/errors"
-	dto "github.com/prometheus/client_model/go"
 	"github.com/spf13/afero"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
 	mount "k8s.io/mount-utils"
 )
 
-func NewAppVolumePublisher(fs afero.Afero, mounter mount.Interface, db metadata.Access, path metadata.PathResolver) csivolumes.Publisher {
-	return &AppVolumePublisher{
+func NewPublisher(fs afero.Afero, mounter mount.Interface, path metadata.PathResolver) csivolumes.Publisher {
+	return &Publisher{
 		fs:      fs,
 		mounter: mounter,
-		db:      db,
 		path:    path,
+		time:    timeprovider.New(),
 	}
 }
 
-type AppVolumePublisher struct {
+type Publisher struct {
 	fs      afero.Afero
 	mounter mount.Interface
-	db      metadata.Access
+	time    *timeprovider.Provider
 	path    metadata.PathResolver
 }
 
-func (publisher *AppVolumePublisher) PublishVolume(ctx context.Context, volumeCfg *csivolumes.VolumeConfig) (*csi.NodePublishVolumeResponse, error) {
-	bindCfg, err := csivolumes.NewBindConfig(ctx, publisher.db, volumeCfg)
-	if err != nil {
-		return nil, err
-	}
-
-	hasTooManyAttempts, err := publisher.hasTooManyMountAttempts(ctx, bindCfg, volumeCfg)
-	if err != nil {
-		return nil, err
-	}
-
-	if hasTooManyAttempts {
+func (pub *Publisher) PublishVolume(ctx context.Context, volumeCfg *csivolumes.VolumeConfig) (*csi.NodePublishVolumeResponse, error) {
+	if pub.hasRetryLimitReached(volumeCfg) {
 		log.Info("reached max mount attempts for pod, attaching dummy volume, monitoring disabled", "pod", volumeCfg.PodName)
 
 		return &csi.NodePublishVolumeResponse{}, nil
 	}
 
-	if !bindCfg.IsArchiveAvailable() {
+	if !pub.isArchiveAvailable(volumeCfg) {
 		return nil, status.Error(
 			codes.Unavailable,
-			"version or digest is not yet set, csi-provisioner hasn't finished setup yet for tenant: "+bindCfg.TenantUUID,
+			"version or digest is not yet set, csi-provisioner hasn't finished setup yet for DynaKube: "+volumeCfg.DynakubeName,
 		)
 	}
 
-	if err := publisher.ensureMountSteps(ctx, bindCfg, volumeCfg); err != nil {
-		return nil, err
+	if err := pub.mountOneAgent(volumeCfg); err != nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("failed to mount oneagent volume: %s", err))
 	}
 
-	agentsVersionsMetric.WithLabelValues(bindCfg.MetricVersionLabel()).Inc()
```

Comment:
you don;'t want to have agentsVersionsMetric anymore?

I checked EPIC, and IMO it would be great if we add somewhere new metrics, so we can monitor it better.

Reply:
I agree, we can iterate on it.